### PR TITLE
Better `Image\Focus` class

### PR DESCRIPTION
**Pull request by distantnative**
## This PR …

Trying to do a better job than the `Focus` class. Logic to focus crop is as follows:
Screenshot 2024-01-16 at 13 40 29

### TODO
- Check to see if the following issues are solved:
  - [x] https://github.com/getkirby/kirby/issues/5855
  - [x] https://github.com/getkirby/kirby/issues/5946
- [x] Bring back `ratio` helper to avoid division by 0

## Ready?

- [ ] Unit tests for fixed bug/feature
- [ ] In-code documentation (wherever needed)
- [ ] Tests and checks all pass

### For review team

- [ ] Add changes to release notes draft in Notion
- [ ] Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)

### CODE REVIEW

File: src/Image/Darkroom/GdLib.php [hunk 1]
```diff
@@ -62,34 +64,12 @@ protected function resize(SimpleImage $image, array $options): SimpleImage
...
-			return $image->thumbnail($options['width'], $options['height']);
-		}
+		// first crop the image in the right ratio
+		$crop  = Focus::fromDarkroomOptions($options);
+		$image = $image->crop($crop->x1, $crop->y1, $crop->x2, $crop->y2);
 
-		// normal crop with crop anchor
-		return $image->thumbnail(
-			$options['width'],
-			$options['height'] ?? $options['width'],
```
Comment (distantnative) [inline src/Image/Darkroom/GdLib.php:91]: @bastianallgeier I am wondering about this old fallback... is it needed? Cause a few lines up we don|t use one for the no-crop variant. And `Darkroom::preprocess()`, which is providing the options, calls `Dimensions::height()` which always returns an integer.

Comment (bastianallgeier) [reply src/Image/Darkroom/GdLib.php:91]: The idea behind using the height as fallback is to create a square crop if no height is specified.

Comment (distantnative) [reply src/Image/Darkroom/GdLib.php:91]: But based on my description, do you see any possibility how that key could not exist? I think there is just no way

Comment (bastianallgeier) [reply src/Image/Darkroom/GdLib.php:91]: Sorry, now I understand what you mean. You are right. This is no longer needed here.

**Comment by distantnative** [comment]
Currently, this is only making some code nicer but not fixing any of the mentioned issues. Closing this for now.

**Pull request closed without merging**